### ISSUE TITLE
type check fail when using Object.freeze

### ISSUE THREAD

**Issue by troy351** [Microsoft/TypeScript#30609]
**TypeScript Version:**  3.4.0-dev.201xxxxx

**Search Terms:** Object.freeze

**Code**

```ts
// A *self-contained* demonstration of the problem follows...
// Test this by running `tsc` on the command-line, rather than through another build tool such as Gulp, Webpack, etc.
const foo: { [key: number]: string } = {
  1: 'val1',
  a: 'val2',
}

const bar: Readonly<{ [key: number]: string }> = Object.freeze({
  1: 'val1',
  a: 'val2',
})
```

**Expected behavior:** both two should show error

**Actual behavior:** only the first one shows error

**Playground Link:** https://www.typescriptlang.org/play/index.html#src=const%20foo%3A%20%7B%20%5Bkey%3A%20number%5D%3A%20string%20%7D%20%3D%20%7B%0D%0A%20%201%3A%20'val1'%2C%0D%0A%20%20a%3A%20'val2'%2C%0D%0A%7D%0D%0A%0D%0Aconst%20bar%3A%20Readonly%3C%7B%20%5Bkey%3A%20number%5D%3A%20string%20%7D%3E%20%3D%20Object.freeze(%7B%0D%0A%20%201%3A%20'val1'%2C%0D%0A%20%20a%3A%20'val2'%2C%0D%0A%7D)

**Related Issues:**

**Comment by dragomirtitian** [comment]
An smaller example of the same issue:

```ts
const foo: { [key: number]: string } = {
  1: 'val1',
  a: 'val2',
}

const obj = {
  1: 'val1',
  a: 'val2',
}
const bar: { [key: number]: string } = obj
```
The error in the first case is triggered by excess property checks (EPC). EPC only works when the object literal is directly assign in a location of a specific type. This is why you get an error in the first case when you assign the object literal directly to the reference. 

In the second case the call to `Object.freez` is checked independently, and the result is typed as `Readonly<{ 1: string , a: string }>` which is assignable to  `Readonly<{ [key: number]: string }>` (the index signature just constrains the object to have all number keys of type string, not that there aren't any extra properties).

One workaround is to specify the type parameter on `freez`, that way you will trigger EPC again on the argument to the function 

```ts
const bar = Object.freeze<{ [key: number]: string }>({
  1: 'val1',
  a: 'val2', // error now
})

```

**Comment by troy351** [comment]
@dragomirtitian thanks for your explanation
